### ISSUE TITLE
Gram Matrix Computation for NTKs

### ISSUE THREAD

**Issue by KonstiNik** [tfjgeorge/nngeometry#94]
Hi,

I am currently working with your NNGeometry library using the gram matrix calculator for NKT computations. 
In the source code (if I got this correctly), I've seen that you compute the `Jacobian` w.r.t. the data rather than the parameters. 
Is that true? If so, can you comment on why and how you do an NTK computation like that?

A very brief script also shows that there are differences of your computation and manual NTK implementations (taken from [here](https://pytorch.org/functorch/stable/notebooks/neural_tangent_kernels.html)).

Cheers, KonstiNik

```python

from nngeometry.generator import Jacobian
from nngeometry.object import FMatDense
from torch.utils.data import TensorDataset

import torch
import torch.nn as nn
from functorch import make_functional
device = 'cuda'


class FC(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc = nn.Linear(3, 3)
        self.fc_out = nn.Linear(3, 1)
        
    def forward(self, x):
        x = self.fc(x)
        x = torch.relu(x)
        x = self.fc(x)
        x = torch.relu(x)
        x = self.fc_out(x)
        return x
    

x = torch.arange(21) - 10
x = x.view(7, 3).float()
x = x.to(device)


# Compute NTK manually

net = FC().to(device)
fnet, params = make_functional(net)

def fnet_single(params, x):
    return fnet(params, x.unsqueeze(0)).squeeze(0)

def empirical_ntk_jacobian_contraction(fnet_single, params, x1, x2):
    # Compute J(x1)
    jac1 = torch.vmap(torch.func.jacrev(fnet_single), (None, 0))(params, x1)
    jac1 = [j.flatten(2) for j in jac1]
    
    # Compute J(x2)
    jac2 = torch.vmap(torch.func.jacrev(fnet_single), (None, 0))(params, x2)
    jac2 = [j.flatten(2) for j in jac2]
    
    # Compute J(x1) @ J(x2).T
    result = torch.stack([torch.einsum('Naf,Mbf->NMab', j1, j2) for j1, j2 in zip(jac1, jac2)])
    result = result.sum(0)
    return result

# Compute twice for comparison
ntk_man = empirical_ntk_jacobian_contraction(fnet_single, params, x, x)


# With nngeometry

def compute_ntk(net, batch):
    generator = Jacobian(model=net, n_output=1)
    K = FMatDense(generator, examples=batch)
    K_torch = K.get_dense_tensor()
    return K_torch

loader = torch.utils.data.DataLoader(TensorDataset(x), batch_size=20)
ntk_nng = compute_ntk(net, loader)

ntk_nng = ntk_nng.swapaxes(0, 1).swapaxes(1, 3) # Brining to the same format


assert torch.allclose(ntk_man, ntk_nng, atol=1e-5)

```

**Comment by tfjgeorge** [comment]
Hi, "Jacobian" is a misnomer for the object that generates the components
to populate FMat and PMat objects. I am planning on changing it soon.

With regards to your code snippet, I will have a look when I have some
time, which should happen tomorrow or Monday :-)

Best

**Comment by tfjgeorge** [comment]
Ok now I get it: in your code snippet, you apply your linear layer twice. I did not have this use case in mind so it is currently a bug, and should at least raise an Exception.

The magic in nngeometry is that you do not require to compute the whole jacobian then apply contraction. Instead the contraction is performed on the fly while the gradient is backpropagated. This way, there is no need to allocate terabytes of memory when working with overparametrized nets. But once again, it currently only works when parameters are not shared.

**Comment by KonstiNik** [comment]
Thanks for clarifying that so quickly! This is very helpful to know. 

Could you comment on the role of the grad that is taken w.r.t. input data in the Jacobian? 
https://github.com/tfjgeorge/nngeometry/blob/6c414de8bca2f1680ba27d6df4ba5b1d8ebc104c/nngeometry/generator/jacobian/jacobian.py#L387
In the case of language models, the input can be ints, for which the NTK computation doesn't work.

**Comment by tfjgeorge** [comment]
The purpose of the call to grad w.r.t. inputs is only to trigger backpropagation all the way down to the input, then actual per-example gradients are computed manually in hooks (see the hooks in generator/jacobian/jacobian and the actual gradient computations in generator/jacobian/grads).

I am aware that this does not work with embedding layers yet because of the integer inputs that cannot be differentiated as you correctly found out, so this is on my TODO list but I can't provide an ETA for that, sorry.

Best

**Comment by KonstiNik** [comment]
Thanks for the clarification. Appreciate your work on this!

Cheers

**Comment by tfjgeorge** [comment]
Hi, you might be interested to know that NNGeometry now supports (vanilla) embedding layers: https://github.com/tfjgeorge/nngeometry/pull/81

**Comment by KonstiNik** [comment]
Great, that helps a lot! Appreciate you letting me know.
